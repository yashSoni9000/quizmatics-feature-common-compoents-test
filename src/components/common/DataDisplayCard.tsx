import { Typography, Box } from "@mui/material";

import ErrorComponent from "@src/components/ErrorComponent";
import UseGenericStateHooks from "@src/hooks/UseGenericStateHook";
import Skeletons from "./Skeletons";

export interface DataDisplayCardData {
    label: string;
    value: number;
    id: number;
}

export type DataDisplayCardProps = ReturnType<
    typeof UseGenericStateHooks<DataDisplayCardData[]>
> & {
    title: string;
    fieldsCount: number;
    handleOnClick: (id: number) => void;
    updateData: () => void;
    className?: string;
};

export default function DataDisplayCard({
    title,
    data,
    error,
    isLoading,
    fieldsCount,
    className,
    updateData,
    refreshData,
    handleOnClick,
}: DataDisplayCardProps) {
    return (
        <Box
            className={`${className} flex flex-col p-4 gap-4 w-full h-full rounded-lg`}
            sx={{
                backgroundColor: "primary.main",
                color: "primary.contrastText",
            }}
        >
            <Typography
                className="mx-auto text-lg md:text-3xl text-center"
                sx={{
                    color: "primary.contrastText",
                }}
            >
                {title}
            </Typography>

            <ErrorComponent
                isLoading={isLoading}
                error={error}
                refreshComponent={() => refreshData(updateData)}
                className="rounded-lg"
                sx={{ backgroundColor: "primary.light" }}
            >
                <Box className="gap-4 flex flex-1 w-full">
                    {isLoading ? (
                        <Skeletons
                            count={fieldsCount}
                            className="min-h-[5.5rem] scale-100 flex-1"
                        />
                    ) : (
                        data.map((item) => {
                            const { id, label, value } = item;
                            return (
                                <Box
                                    key={id}
                                    className={`flex flex-col flex-1 justify-between mx-auto p-2 min-h-[5.5rem] scale-100 
                                                    rounded-lg transition-colors duration-300 hover:cursor-pointer`}
                                    onClick={() => handleOnClick(id)}
                                    sx={{
                                        backgroundColor: "primary.light",
                                        "> p": {
                                            color: "primary.contrastText",
                                        },
                                        ":hover": {
                                            backgroundColor: "primary.dark",
                                        },
                                    }}
                                >
                                    <Typography className="text-center">
                                        {label}
                                    </Typography>
                                    <Typography className="text-xl mt-2 text-center font-bold">
                                        {value}
                                    </Typography>
                                </Box>
                            );
                        })
                    )}
                </Box>
            </ErrorComponent>
        </Box>
    );
}
