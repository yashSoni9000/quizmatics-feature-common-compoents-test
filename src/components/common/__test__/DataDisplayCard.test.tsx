import { fireEvent, render, renderHook, screen } from "@testing-library/react";
import "@testing-library/jest-dom";

import DataDisplayCard, {
    DataDisplayCardData,
    DataDisplayCardProps,
} from "../DataDisplayCard";
import UseGenericStateHooks from "@src/hooks/UseGenericStateHook";

let onClickMock: ReturnType<typeof jest.fn>;
let dataDisplayCardData: ReturnType<
    typeof UseGenericStateHooks<DataDisplayCardData[]>
>;
const mockData = { id: 1, label: "Label", value: 20 };

function DataDisplayCardComponent({ ...props }: Partial<DataDisplayCardProps>) {
    return (
        <DataDisplayCard
            {...dataDisplayCardData}
            title="title"
            fieldsCount={1}
            handleOnClick={onClickMock}
            updateData={jest.fn()}
            isLoading={false}
            {...props}
        />
    );
}

jest.useFakeTimers();

beforeEach(() => {
    onClickMock = jest.fn();
    dataDisplayCardData = renderHook(() =>
        UseGenericStateHooks<DataDisplayCardData[]>([mockData]),
    ).result.current;
});

describe("Data Display Card Test", () => {
    it("snapshot test", () => {
        const { asFragment } = render(<DataDisplayCardComponent />);

        const label = screen.getByText(mockData.label);
        expect(label).toBeVisible();

        expect(asFragment()).toMatchSnapshot();
    });

    it("Skeletons are visible when card is loading", () => {
        render(<DataDisplayCardComponent isLoading={true} />);

        const skeleton = document.querySelector(".MuiSkeleton-root");
        expect(skeleton).toBeVisible();
    });

    it("Clicking on card button invokes handleOnChange", () => {
        render(<DataDisplayCardComponent />);

        const cardButton = screen.getByText(mockData.label);
        fireEvent.click(cardButton);

        expect(onClickMock).toHaveBeenCalledWith(mockData.id);
    });

    it("DataDisplayCard simulation from error state to non-error state", async () => {
        let loading = false,
            error = true;

        function refreshComponentMock() {
            error = false;
            loading = true;

            setTimeout(() => {
                loading = false;
            }, 1000);
        }

        function GenericDataDisplayCardComponent() {
            return (
                <DataDisplayCardComponent
                    isLoading={loading}
                    error={error}
                    refreshData={refreshComponentMock}
                />
            );
        }

        function rerenderComponent() {
            rerender(<GenericDataDisplayCardComponent />);
        }

        const { rerender, asFragment } = render(
            <GenericDataDisplayCardComponent />,
        );

        const retryButton = screen.getByText("Retry").parentElement;
        if (!retryButton) throw new Error();

        expect(retryButton).toBeVisible();
        fireEvent.click(retryButton);

        rerenderComponent();

        const skeleton = document.querySelector(".MuiSkeleton-root");
        expect(skeleton).toBeVisible();

        jest.advanceTimersByTime(1000);
        rerenderComponent();

        const label = screen.getByText(mockData.label);
        expect(label).toBeVisible();

        expect(asFragment()).toMatchSnapshot();
    });
});
