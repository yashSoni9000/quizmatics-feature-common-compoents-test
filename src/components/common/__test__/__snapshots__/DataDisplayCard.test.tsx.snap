// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data Display Card Test DataDisplayCard simulation from error state to non-error state 1`] = `
<DocumentFragment>
  <div
    class="undefined flex flex-col p-4 gap-4 w-full h-full rounded-lg MuiBox-root css-1u8r73e"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 mx-auto text-lg md:text-3xl text-center css-4peds9-MuiTypography-root"
    >
      title
    </p>
    <div
      class="gap-4 flex flex-1 w-full MuiBox-root css-0"
    >
      <div
        class="flex flex-col flex-1 justify-between mx-auto p-2 min-h-[5.5rem] scale-100 
                                                    rounded-lg transition-colors duration-300 hover:cursor-pointer MuiBox-root css-1aytai0"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 text-center css-ahj2mt-MuiTypography-root"
        >
          Label
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1 text-xl mt-2 text-center font-bold css-ahj2mt-MuiTypography-root"
        >
          20
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Data Display Card Test snapshot test 1`] = `
<DocumentFragment>
  <div
    class="undefined flex flex-col p-4 gap-4 w-full h-full rounded-lg MuiBox-root css-1u8r73e"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 mx-auto text-lg md:text-3xl text-center css-4peds9-MuiTypography-root"
    >
      title
    </p>
    <div
      class="gap-4 flex flex-1 w-full MuiBox-root css-0"
    >
      <div
        class="flex flex-col flex-1 justify-between mx-auto p-2 min-h-[5.5rem] scale-100 
                                                    rounded-lg transition-colors duration-300 hover:cursor-pointer MuiBox-root css-1aytai0"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 text-center css-ahj2mt-MuiTypography-root"
        >
          Label
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1 text-xl mt-2 text-center font-bold css-ahj2mt-MuiTypography-root"
        >
          20
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
